### ISSUE TITLE
List of Pathfinder issues

### ISSUE THREAD

**Issue by SmallJoker** [minetest/minetest#8642]
##### Performance
 - [x] Occasional excessive work load #8584
 - [ ] Alpine Linux: stack overflow #7132

##### Bugs
 - [x] Jump/fall distances cause glitches #4198
 - [ ] Broken on non-cubic nodes #4192
 - [x] Only "walks" through air #1701
 - [x] Too complicated A* paths #9331

New pathfinder discussion: https://github.com/minetest/minetest/issues/5479
More requested features: https://github.com/minetest/minetest/issues/2614

**Comment by thomasrudin** [comment]
What about a warning in the `lua_api.txt` to discourage modders from using the pathfinder (until a better solution/implementation is found)

**Comment by paramat** [comment]
It is not that bad, it still functions reasonably well. It is essential and there is no alternative.

**Comment by thomasrudin** [comment]
> It is not that bad, it still functions reasonably well. It is essential and there is no alternative.

Heh, that message was half a year ago, meanwhile i resorted to this:
https://github.com/pandorabox-io/pandorabox_custom/blob/master/hacks.lua

```lua
minetest.find_path = function()
	return nil
end
```

This works very well with all the `mob_redo` variants and the mobs themselves don't seem to react very differently :smile: 

Usage in the `mobs_redo` mod where it handles nil return values:
https://notabug.org/TenPlus1/mobs_redo/src/master/api.lua#L1494

**Comment by GreenXenith** [comment]
I had to roll my own Lua-based A* pathfinder for mobs because the Minetest pathfinder doesn't return useful failure data. Returning the reason for failure is very important, aka which endpoint is enclosed causing the failure.

**Comment by rubenwardy** [comment]
> aka which endpoint is enclosed causing the failure.

I don't understand what this means. Do you mean returning the closest node to the destination?

**Comment by GreenXenith** [comment]
> I don't understand what this means. Do you mean returning the closest node to the destination?

No. I quite literally mean "which endpoint caused the failure". If a path fails because one endpoint is enclosed, we need to know which one was enclosed.

**Comment by rubenwardy** [comment]
> No. I quite literally mean "which endpoint caused the failure". If a path fails because one endpoint is enclosed, we need to know which one was enclosed.

This doesn't make sense in terms of pathfinder algorithms. So, by endpoint do you mean start or from, and by enclosed, do you mean: not open to the open environment?

**Comment by GreenXenith** [comment]
> To confirm: by endpoint do you mean start or from

Yes. I mean "which point couldn't couldn't find its way out: start or end (or both?)?".

> by enclosed, do you mean not open to the open environment?

If it is enclosed, it means it failed to connect to the other point for some reason or another.

The pathfind function returns nil if a path couldn't be found. In a lot of cases, one needs to know *why* and *where* it failed.

**Comment by Wuzzy2** [comment]
Add #9331 to the list.

**Comment by thomasrudin** [comment]
> I had to roll my own Lua-based A* pathfinder for mobs

@GreenXenith can you share that implementation by any chance?

**Comment by GreenXenith** [comment]
> can you share that implementation by any chance?

@thomasrudin sure. I implemented it for [the Infinite IKEA Staff](https://gitlab.com/benrob0329/ikea/blob/master/mods/ikea_staff/pathfind.lua) ([relevant code](https://gitlab.com/benrob0329/ikea/blob/master/mods/ikea_staff/pathfind.lua#L70-204)). No, it is not the cleanest, but it works (mostly .. there are a few odd movement bugs that I have yet to trace .. could be the pathfinder but I'd like to not think otherwise).

**Comment by rubenwardy** [comment]
> If it is enclosed, it means it failed to connect to the other point for some reason or another.

This doesn't really make sense as part of a pathfinder, and "is it connected to the outside world" is rather arbitrary to decide - what if it's in a valley?

I briefly looked through your code, does it do pathfinding in both directions?

**Comment by GreenXenith** [comment]
> what if it's in a valley?

Terrain has no meaning. If point A is in the valley and point B is outside of it, and there is a valid path to it, everything is fine.

> if A is enclosed and B is not, then it is possible to get from B to A but not from A to B

If A is enclosed, _nothing_ will get to it. If B is enclosed, _nothing_ will get to it. I want to know which ends, if any, are enclosed.

> does it do pathfinding in both directions?

That is how A* works. Pathfinding is started from both ends in a circular radius until the "circles" touch, at which point a valid path is found.
If the circles expand past an arbitrary radius without touching, the points are just too far apart.
If a path dies before meeting, it means that end is enclosed. This is *very* useful information as to why a path failed. In the case of mobs, it means "Am I trapped?" or "Can I not get to my destination?". Minetest does not give us this information.

**Comment by rubenwardy** [comment]
That is not how A* works. A* uses a heuristic to go from the start to the end. You don't expand from the end

> Terrain has no meaning. If point A is in the valley and point B is outside of it, and there is a valid path to it, everything is fine.

By this I meant a large section of land which you cannot leave due to high walls, which is still outside

**Comment by rubenwardy** [comment]
https://www.redblobgames.com/pathfinding/a-star/introduction.html

**Comment by rubenwardy** [comment]
Your definition of "enclosed" needs to be more specific in order to be useful.

For example: are there any valid paths out of an N-radius sphere from position X?

**Comment by GreenXenith** [comment]
Ok, you are correct as to it not being A*. I was thinking A* because that is originally how it was supposed to be implemented, but I deviated. My pathfinder is actually a breadth-first search (sorry @thomasrudin). Not entirely important, however.

The key lies in what you added to a previous comment:
> In which case, you should instead run the pathfinder twice, one for each direction

I don't know why I never thought of this. _However_, the only reason this works is because pathfinding is as fast as it is. If my implementation took 750us, it would be faster to use it than run the Minetest pathfinder twice. But I suppose that isn't really worth worrying about.

> Your definition of "enclosed" needs to be more specific in order to be useful.

Not so.
Enclosed: "Surrounded or closed off on all sides."
In other words: Point A or Point B has no valid path to the other (which may or may not be within an arbitrary radius).

**Comment by GreenXenith** [comment]
>> In which case, you should instead run the pathfinder twice, one for each direction
>
> I don't know why I never thought of this

I take this back.
Take the following case where point A is sealed and point B is not (but we don't know this yet):
If I run a pathfinder from point A to point B, it will return nil.
If I run a pathfinder from point B to point A, it will return nil.
What have I learned? There is no path from point A to point B. Is point A enclosed or is point B enclosed? The pathfinder doesn't tell me.

Running it twice won't get you different results. The points are still the same points, the only difference is the direction you search (which in the end influences nothing).

**Comment by rubenwardy** [comment]
That is not a useful definition due to its ambiguous phrasing. A better definition would be: a point A is enclosed if there is no path to point B. However, there may be a path from point B to A

**Comment by rubenwardy** [comment]
Ok, we've decided on an accurate definition of how enclosed should be determined: a point is enclosed if it exhausted all possible nodes before reaching the other node, rather than reaching distance/time limits

**Comment by Wuzzy2** [comment]
Recently fixed issues:

* Jump/fall distances cause glitches #4198
* Only "walks" through air #1701
* Too complicated A* paths #9331

**Comment by rubenwardy** [comment]
this is what a label should be used for
